File: Programmers/Level_3/단어변환.py
from collections import deque

def solution(begin, target, words):
    answer = cnt = 0
    visited = [False] * len(words)

    def bfs():
        q = deque()
        q.append((begin, cnt))
        
        while q:
            tmp, idx = q.popleft()
            if tmp == target:
                nonlocal answer
                answer = idx
                break

            for i, word in enumerate(words):
                if visited[i] == False:
                    # 다른 글자 개수가 1개이면 그 단어를 큐에 넣고 카운트 1 증가 
                    if [c1 == c2 for c1, c2 in zip(tmp, word)].count(False) == 1:
                        visited[i] = True 
                        q.append((word, idx+1))
                    
    bfs()
    return answer

solution("hit", "cog", ["hot", "dot", "dog", "lot", "log"])

File: Programmers/Level_3/Swift_Code/Swift_Code/단어변환.swift
//
//  단어변환.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/05/25.
//

/*
 https://programmers.co.kr/learn/courses/30/lessons/43163?language=swift
 
 */

import Foundation

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    var answer = Int.max
    var visited = [Bool](repeating: false, count: words.count)
    
    if !words.contains(target) {
        return 0
    }
    
    func diffCount(_ a: String, _ b: String) -> Int {
        return zip(a, b).filter { $0.0 != $0.1 }.count
    }
    
    func dfs(_ cnt: Int, _ tmp: String) {
        if cnt >= answer {
            return
        }
        
        if tmp == target {
            answer = cnt
            return
        }
        
        for i in 0..<words.count {
            if !visited[i] && diffCount(tmp, words[i]) < 2 {
                visited[i] = true
                dfs(cnt + 1, words[i])
                visited[i] = false
            }
        }
    }
    
    dfs(0, begin)
    return answer
}

// print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))
